python/ray/tune/schedulers/pb2.py

```diff
@@ -0,0 +1,322 @@
+import pandas as pd
+import GPy
+
+from ray.tune.schedulers import PopulationBasedTraining
+from ray.tune.schedulers.pb2_utils import normalize, optimize_acq, \
+    select_length, \
+    standardize, TV_SquaredExp
+
+logger = logging.getLogger(__name__)
+
+
+## select config, given X and y.
+def select_config(Xraw, yraw, current, newpoint, bounds, num_f, num):
+
+    length = select_length(Xraw, yraw, current, newpoint, bounds, num_f, num)
+
+    Xraw = Xraw[-length:, :]
+    yraw = yraw[-length:]
+
+    base_vals = np.array(list(bounds.values())).T
+    oldpoints = Xraw[:, :num_f]
+    old_lims = np.concatenate((np.max(oldpoints, axis=0),
+                               np.min(oldpoints, axis=0))).reshape(
+                                   2, oldpoints.shape[1])
+    limits = np.concatenate((old_lims, base_vals), axis=1)
+
+    X = normalize(Xraw, limits)
+    y = standardize(yraw).reshape(yraw.size, 1)
+
+    fixed = normalize(newpoint, oldpoints)
+
+    kernel = TV_SquaredExp(
+        input_dim=X.shape[1], variance=1., lengthscale=1., epsilon=0.1)
+
+    try:
+        m = GPy.models.GPRegression(X, y, kernel)
+    except np.linalg.LinAlgError:
+        # add diagonal ** we would ideally make this something more robust...
+        X += np.eye(X.shape[0]) * 1e-3
+        m = GPy.models.GPRegression(X, y, kernel)
+
+    try:
+        m.optimize(messages=True)
+    except np.linalg.LinAlgError:
+        # add diagonal ** we would ideally make this something more robust...
+        X += np.eye(X.shape[0]) * 1e-3
+        m = GPy.models.GPRegression(X, y, kernel)
+        m.optimize(messages=True)
+
+    m.kern.lengthscale.fix(m.kern.lengthscale.clip(1e-5, 1))
+
+    # m1 is used to reduce redundancy. If there are trials already running (current) we can use this info in the acquisition function
+    if current is None:
+        m1 = deepcopy(m)
+    else:
+        # add the current trials to the dataset
+        padding = np.array([fixed for _ in range(current.shape[0])])
+        current = normalize(current, base_vals)
+        current = np.hstack((padding, current))
+
+        Xnew = np.vstack((X, current))
+        ypad = np.zeros(current.shape[0])
+        ypad = ypad.reshape(-1, 1)
+        ynew = np.vstack((y, ypad))
+
+        #kernel = GPy.kern.RBF(input_dim=X.shape[1], variance=1., lengthscale=1.)
+        kernel = TV_SquaredExp(
+            input_dim=X.shape[1], variance=1., lengthscale=1., epsilon=0.1)
+        m1 = GPy.models.GPRegression(Xnew, ynew, kernel)
+        m1.optimize()
+
+    xt = optimize_acq(UCB, m, m1, fixed, num_f)
+
+    # convert back...
+    xt = xt * (np.max(base_vals, axis=0) - np.min(base_vals, axis=0)) + np.min(
+        base_vals, axis=0)
+
+    xt = xt.astype(np.float32)
+    return (xt)
+
+
+## Outer loop: set up data/bounds, pass into select_config function
+def explore(data, bounds, current, base, old, config):
+
+    data['Trial'] = data['Trial'].astype(str)
+    data.to_csv('old.csv')
```

Comment:
Is it necessary to save the data here? This could be confusing for the user if they see a bunch of csv files and they don't know where it came from.

Reply:
Good catch - this was for debugging during the research phase so it is not necessary here. I have removed it.